modules/hitlnext/src/backend/middleware.ts

```diff
@@ -0,0 +1,191 @@
+
+const debug = DEBUG(MODULE_NAME)
+
+const registerMiddleware = async (bp: typeof sdk, state: StateType) => {
+  const cache = new LRU<string, string>({ max: 1000, maxAge: ms('1 day') })
+  const repository = new Repository(bp, state.timeouts)
+  const realtime = Socket(bp)
+
+  const pipeEvent = async (event: sdk.IO.IncomingEvent, eventDestination: sdk.IO.EventDestination) => {
+    debug.forBot(event.botId, 'Piping event', eventDestination)
+    return bp.events.replyToEvent(eventDestination, [{ type: 'typing', value: 10 }, event.payload])
+  }
+
+  const cacheKey = (a, b) => [a, b].join('.')
+
+  const getHandoff = (botId: string, threadId: string) => {
+    return cache.get(cacheKey(botId, threadId))
+  }
+
+  const cacheHandoff = (botId: string, threadId: string, handoff: IHandoff) => {
+    debug.forBot(botId, 'Caching handoff', { id: handoff.id, threadId })
+    cache.set(cacheKey(botId, threadId), handoff.id)
+  }
+
+  const expireHandoff = (botId: string, threadId: string) => {
+    debug.forBot(botId, 'Expiring handoff', { threadId })
+    cache.del(cacheKey(botId, threadId))
+  }
+
+  const handleIncomingFromUser = async (handoff: IHandoff, event: sdk.IO.IncomingEvent) => {
+    if (handoff.status === 'assigned') {
+      // There only is an agentId & agentThreadId after assignation
+      await pipeEvent(event, {
+        botId: handoff.botId,
+        target: handoff.agentId,
+        threadId: handoff.agentThreadId,
+        channel: 'web'
+      })
+    }
+
+    // At this moment the event isn't persisted yet so an approximate
+    // representation is built and sent to the frontend, which relies on
+    // this to update the handoff's preview and read status.
+    const partialEvent = {
+      event: _.pick(event, ['preview']),
+      success: undefined,
+      threadId: undefined,
+      ..._.pick(event, ['id', 'direction', 'botId', 'channel', 'createdOn', 'threadId'])
+    }
+
+    realtime.sendPayload(event.botId, {
+      resource: 'handoff',
+      type: 'update',
+      id: handoff.id,
+      payload: {
+        ...handoff,
+        userConversation: partialEvent
+      }
+    })
+
+    realtime.sendPayload(event.botId, {
+      resource: 'event',
+      type: 'create',
+      id: null,
+      payload: partialEvent
+    })
+  }
+
+  const handleIncomingFromAgent = async (handoff: IHandoff, event: sdk.IO.IncomingEvent) => {
+    const { botAvatarUrl }: Config = await bp.config.getModuleConfigForBot(MODULE_NAME, event.botId)
+    Object.assign(event.payload, {
+      from: 'agent',
+      botAvatarUrl
+    }) // TODO set avatar url from agent profile if nothing in config
+
+    await pipeEvent(event, {
+      botId: handoff.botId,
+      threadId: handoff.userThreadId,
+      target: handoff.userId,
+      channel: handoff.userChannel
+    })
+
+    await repository.setAgentOnline(event.botId, handoff.agentId, async () => {
+      // By now the agent *should* be offline, but we check nonetheless
```

Comment:
`setAgentOnline` accepts a callback that fires when the agent's session expires. So when the callback is called, the agent *should* be offline when checking in the KVS, but if in the meantime (milliseconds really) the agent has become online again we could be sending incorrect information in the `realtime.sendPayload` call. In short, we "double check" so that the session expiry & callback are *almost* transactional and resilient to an improbable but possible race condition, for the cost 1 sql query.